src/Microsoft.TemplateEngine.TemplateLocalizer.Core/TemplateStringExtractor.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using Microsoft.TemplateEngine.TemplateLocalizer.Core.KeyExtractors;
+using Microsoft.TemplateEngine.TemplateLocalizer.Core.TraversalRules;
+
+namespace Microsoft.TemplateEngine.TemplateLocalizer.Core
+{
+    /// <summary>
+    /// Extracts localizable strings from template.json files.
+    /// </summary>
+    internal sealed class TemplateStringExtractor
+    {
+        private readonly JsonDocument _jsonDocument;
+
+        private readonly IJsonKeyCreator _defaultArrayKeyExtractor = new IndexBasedKeyCreator();
+        private readonly IJsonKeyCreator _defaultObjectKeyExtractor = new PropertyNameBasedKeyCreator();
+
+        private readonly List<TemplateString> _extractedStrings = new List<TemplateString>();
+        private readonly HashSet<string> _extractedStringIds = new HashSet<string>();
+
+        /// <summary>
+        /// The list of rules that define which fields in the template json should be extracted to a templatestrings.json file.
+        /// </summary>
+        private readonly List<TraversalRule> _traversalRules = new List<TraversalRule>()
+        {
+            new StringFilteredTraversalRule(string.Empty).WithChildren(
```

Comment:
This is only needed because the root element has no name. I don't think it will be a bottleneck in any point. Why do you think we should change it? 
There is also an `AllInclusiveTraversalRule` that just accepts anything, but I chose this one to be more explicit with what we want: root name is always empty string.

Reply:
I tried to rename it, but couldn't come up with something that makes sense. The type name itself has nothing to do with root element. These are individual rules and renaming it results in a type where the name doesn't match what it actually does.
I understand that these rules don't read like you would normally read English. But maybe we shouldn't expect them to. I think we are moving away from what this actually does and focusing too much on how it looks.